### DISCUSSION TITLE
Option to set smallest size while resizing

### DISCUSSION THREAD

**Discussion by aried3r** [libvips/libvips#2007] [Ideas]
I'm currently studying the extensive libvips API but haven't found anything there or in the issues to achieve the following:

ImageMagick has a [_"geometry option flag '^' which is used to resize the image based on the smallest fitting dimension."_](https://www.imagemagick.org/Usage/resize/#fill), but I haven't found anything to do this with libvips directly, especially the ruby bindings without calculating the width/height myself (not that this would be a problem).  
I haven't found anything in [resample](https://jcupitt.github.io/libvips/API/current/libvips-resample.html), so I figured this might not be implemented.

Is this currently achievable using libvips? And if not, would it be an option to have it in future versions?

Admittedly I think this might be a rather obscure use case since images like panoramas ([example](https://commons.wikimedia.org/wiki/File:Clocktower_Panorama_20080622_20mb.jpg)) would result in very wide thumbnails. However in my case I have full control over the input images and no such thing can happen.

**Comment by jcupitt** [comment]
You're right, there's no direct API for that. You could do it in Ruby with something like:

```ruby
# set the target size for the smaller axis to this
target_smallest = 80

image = Vips::Image.new_from_file "filename"
infinity = 10000000
target_width = image.width < image.height ? target_smallest : infinity
target_height = image.height < image.width ? target_smallest : infinity
thumb = Vips::Image.thumbnail "filename", target_width, height: target_height
```

The `infinity` is a bit ugly. git master ruby-vips has the constant `Vips::MAX_COORD`, which does the same job.

**Reply by janko** [reply]
For people stumbling across this, for non-square boxes the logic is a bit more complex:

```rb
target_width, target_height = 300, 400 # container dimensions
image = Vips::Image.new_from_file "filename"

image_ratio = Rational(image.width, image.height)
thumbnail_ratio = Rational(target_width, target_height)

if image_ratio > thumbnail_ratio
  target_width = ::Vips::MAX_COORD
else
  target_height = ::Vips::MAX_COORD
end

thumb = Vips::Image.thumbnail "filename", target_width, height: target_height
```

**Comment by aried3r** [comment]
Thanks :)

Is there any chance to have this option built into libvips or would that add too much complexity?

**Comment by jcupitt** [comment]
Let's tag this an an enhancement. It would be a simple addition to `vipsthumbnail.c` at least.
